### PR TITLE
[Jasoet|Mahamuni] Add ability to fetch hosts from yggdrasil

### PR DESCRIPTION
**Pull request by jasoet**

### CODE REVIEW

File: cookbooks/zookeeper/README.md [hunk 1]
```diff
@@ -15,23 +15,6 @@ bundle exec berks vendor cookbooks
 2. Commit and updated `cookbooks` directory
 3. Tag the commit that we want to release with format `<APP-VERSION>-<REVISION>`
 
-## Building Image
```
Comment (giosakti) [inline cookbooks/zookeeper/README.md]: why did you remove this README section?

File: libraries/yggdrasil.rb [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+    end
+
+    raise 'Failed to fetch config from yggdrasil!' unless response.code == '200'
+    response_body = JSON.parse(response.body)
+    raise 'Failed to fetch config from yggdrasil!' unless response_body['success']
+    response_body['data']
+  end
+
+end
+
```
Comment (giosakti) [inline libraries/yggdrasil.rb]: remove unnecessary extra whitespaces

File: libraries/yggdrasil.rb [hunk 3]
```diff
@@ -0,0 +1,26 @@
+class Yggdrasil
+  def initialize(host,api_version,token)
+      @host = host
+      @api_version = api_version
+      @token = token
+  end
+
+  def fetch_configs(namespace, overrides)
+    uri = URI("http://#{@host}:80/#{@api_version}/configurations/#{namespace}/latest?q=#{overrides}")
```
Comment (giosakti) [inline libraries/yggdrasil.rb]: parameterize port

File: recipes/search.rb [hunk 4]
```diff
@@ -11,6 +12,19 @@
...
+
+  yggdrasil = Yggdrasil.new(yggdrasil_host,yggdrasil_api_version,yggdrasil_token)
+  yggdrasil_config = yggdrasil.fetch_configs(yggdrasil_namespace,yggdrasil_overrides)
+
+  node.run_state[cookbook_name]['hosts'] = yggdrasil_config["zookeeper_hosts"]
+else
+  node.run_state[cookbook_name]['hosts'] = node[cookbook_name]['hosts']
+end
+
 node.run_state[cookbook_name]['my_id'] = node[cookbook_name]['my_id']
```
Comment (giosakti) [inline recipes/search.rb]: does yggdrasil store my_id ? (sequentially generated number for identifying member of zookeeper cluster)

File: spec/unit/recipes/default_spec.rb [hunk 5]
```diff
@@ -5,18 +5,21 @@
...
 
 require 'spec_helper'
+require_relative '../../../libraries/yggdrasil'
 
 describe 'zookeeper::default' do
   context 'When all attributes are default, on Ubuntu 16.04' do
     let(:chef_run) do
       # for a complete list of available platforms and versions see:
       # https://github.com/customink/fauxhai/blob/master/PLATFORMS.md
+      mock_result = {zookeeper_hosts: ['zookeeper-01-ubuntu-1804.vagrantup.com', 'zookeeper-02-ubuntu-1804.vagrantup.com']}
```
Comment (giosakti) [inline spec/unit/recipes/default_spec.rb]: is this test working? you haven't specified yggdrasil in kitchen to be `true`
please create additional scenario in kitchen to test when yggdrasil is true

File: cookbooks/zookeeper/README.md (hunk 1 above)
Comment (jasoet) [reply cookbooks/zookeeper/README.md]: I do not, but somehow it removed. I will revert it back.

File: libraries/yggdrasil.rb (hunk 2 above)
Comment (jasoet) [reply libraries/yggdrasil.rb]: Sure

File: recipes/search.rb (hunk 4 above)
Comment (jasoet) [reply recipes/search.rb]: No, It must be supplied from node attribute

**Comment by giosakti** [comment]
so what to do now regarding intermittent failures that can block our CI/CD?
any ideas?

also because the provisioner in Vagrantfile is ignored, maybe we can create a cx script instead?

**Comment by jasoet** [comment]
@giosakti I am experimenting on using combined Vagrantfile on kitchen-vagrant https://github.com/test-kitchen/kitchen-vagrant#-vagrantfiles. With this, we should able to use the kitchen to do a proper test by adding provisioner.

I will also create the cx-script for the provisioner.

**Comment by jasoet** [comment]
@giosakti It is now using combined Vagrantfile feature on kitchen-vagrant. It will bootstrap the Mock Server when do kitchen create.

**Comment by jasoet** [comment]
@giosakti please check this PR, this should be final.

**Comment by giosakti** [comment]
this pattern is okay now, however I just realized another thing.

@jasoet will @bimoZX work on this now? I suggest that we create another cookbook called `yggdrasil-cookbook` and then move the client implementation there.. so that zookeeper cookbook can be 'loosely-coupled' with yggdrasil-related functionalities

**Comment by giosakti** [comment]
Similar with Barito's [consul-cookbook](https://github.com/BaritoLog/consul-cookbook) see [register service](https://github.com/BaritoLog/consul-cookbook/blob/master/resources/register_service.rb) recipe.

**Comment by jasoet** [comment]
@giosakti that will be better approach. I will coordinate with @BimoZX about this.

**Comment by giosakti** [comment]
Closed in favor of #5
